fix: update `fromIndex` handling in `blas/ext/base/ndarray/dlast-index-of`

headlessNode opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: na
  - task: lint_repl_help status: na
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: na
  - task: lint_license_headers status: passed ---

Progresses #2656.

## Description

> What is the purpose of this pull request?

This pull request:

-   fix `fromIndex` handling in `blas/ext/base/ndarray/dlast-index-of`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2656

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
refactor: fix fromIndex handling

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
@kgryte I think only this function needed the change, `blas/ext/base/dlast-index-of` works as expected. And coming from top-level `fromIndex` will get resolved here.

kgryte commented:
@headlessNode We'll also need to update `slast-index-of`, correct? https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/blas/ext/base/slast-index-of

kgryte committed:
docs: update example to preserve original intent

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/dlast-index-of/lib/main.js:
```diff
@@ -78,14 +78,11 @@ function dlastIndexOf( arrays ) {
 	} else if ( fromIndex >= N ) {
 		return -1;
```
@headlessNode This behavior does not match `Array.prototype.lastIndexOf` (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/lastIndexOf#fromindex). Namely, when `fromIndex` exceeds the last index `N-1`, the `fromIndex` is clamped to the last index. We should probably follow similarly here. We should also check that `index-of` semantics more or less align with `Array.prototype.indexOf`.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/dlast-index-of/lib/main.js:
```diff
@@ -78,14 +78,11 @@ function dlastIndexOf( arrays ) {
 	} else if ( fromIndex >= N ) {
```
L76 also does not align with `Array.prototype.lastIndexOf`. Looks like you just copied these blocks from `index-of`, where those blocks are correct. Here, they are not correct. You need to fix the bounds checking behavior.

kgryte reviewed (changes requested):
Other than the bounds checking behavior, this is shaping up. Once the bounds behavior is addressed, you should double-check that the tests are correct.

headlessNode committed:
fix: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/ndarray/dlast-index-of/lib/main.js:
```diff
@@ -70,22 +70,20 @@ function dlastIndexOf( arrays ) {
-		if ( fromIndex < 0 ) {
-			fromIndex = 0;
+	if (fromIndex < 0) {
+		if (fromIndex >= -N) {
```
```suggestion
	if ( fromIndex < 0 ) {
		if ( fromIndex >= -N ) {
```

kgryte committed:
style: fix spacing

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
refactor: reduce branch complexity

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
refactor: inline expressions to reduce number of variables on stack

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM after some minor clean-up.

kgryte commented:
@headlessNode I believe we now need to update `slast-index-of`, correct?

headlessNode commented:
@kgryte Yep.

kgryte commented:
@headlessNode ...and `glast-index-of`. 😅
